### PR TITLE
docs: Add article on how to view epoch metrics

### PR DESCRIPTION
**Pull request by tara-det-ai**
How to view epoch-based metrics in the WebUI.

## Dependencies

#7542 

## Functionality

Before the implementation of #7542, users needed to select the Compare button from their list of experiments in order to enable the x-axis dropdown. Now, users can run an experiment and report epoch-based metrics, and interact with the x-axis dropdown from the experiment overview.

how to view epoch metrics in xaxis selector

## Checklist

- [x] Code has been reviewed
- [x] Changes have been manually QA'd by tech writing
- [x] Steps are reproducible by tech writing
- [x] Content has been reviewed by community team
- [x] Content meets style guidelines

### CODE REVIEW

**Review by ashtonG — APPROVED** [review]

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+.. code:: python
+
+   # NEW: Report epoch-based training metrics to Determined
+   # master via core_context.
+   # Index by (batch_idx + 1) * (epoch-1) * len(train_loader)
+   # to continuously plot loss on one graph for consecutive
+   # epochs.
+   core_context.train.report_training_metrics(
+       steps_completed=batches_completed + epoch_idx * len(train_loader),
+       metrics={"train_loss": loss.item(), "epoch": epoch_idx},
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:119]: Is all you need to do to get epoch indices to pass `"epoch": epoch_idx`?

I feel like this is several pages of docs to explain just that.

Is there somewhere we can say that in a much, much more concise way?

Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst:119]: per our discussion, this is resolved

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+headers accordingly.
+
+To follow along, use the ``model_def_metrics.py`` script and its accompanying ``metrics.yaml``
+experiment configuration file.
+
+We'll start by importing Determined:
+
+.. code:: bash
+
+   python3
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:94]: Drop the `python3` code snippet.

It doesn't look like you say this anywhere, this snippet is implying that somebody is running the code directly from a python3 interpreter.

That is something which few people will do for anything more than very trivial code examples, and our end users are far more likely to do it in a notebook than in a `python3` interpreter.

Also, in general coding tutorials assume the reader knows how to run code in their chosen environment, so I would say it's non-standard to include this detail anyway.

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+.. code:: python
+
+   import determined as det
+
+Step 2.1: Verify the Main Loop Contains ``core.Context``
+========================================================
+
+We'll need a ``core.Context`` object for interacting with the master. Verify the __main__loop in
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:103]: `__main__` needs backticks, and it's not a loop.

I usually refer to it as "the `if __name__ == __main__` block" because technically `__main__` refers to the entire script you pass to the python3 interpreter, not just this one block of code.

Comment (rb-determined-ai) [reply docs/articles/viewing-epoch-based-metrics.rst:103]: side note, why are they "verifying" something that is already true about a script they just downloaded from us?

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+include ``epoch_idx`` as a metric. This allows Determined to keep track of the specific epoch for
+which training loss is being reported:
+
+.. code:: python
+
+   # NEW: Report epoch-based training metrics to Determined
+   # master via core_context.
+   # Index by (batch_idx + 1) * (epoch-1) * len(train_loader)
+   # to continuously plot loss on one graph for consecutive
+   # epochs.
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:131]: this `# NEW` comment is really specific to the `model_def.py` -> `model_def_metrics.py` transition, and is pretty confusing in the context of this document I think.

Can it just be omitted?

Also, this comment applies to all the `# New` comments in code snippets throughout this example, I think.  Looks like they all have one.

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+   )
+
+Now that we've reported an epoch value, **Epoch** will be an available option for the X-Axis when we
+view our metric data graph in the WebUI.
+
+Step 2.3: Verify the Test Method
+================================
+
+We'll need a ``test()`` function to evaluate the trained model on the test/validation data for the
+current epoch.
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:158]: This step doesn't make much sense, because the presence of a step suggest that there's going to be some code changes, but it looks to me like the code calling `test()` is already correct?

Comment (rb-determined-ai) [reply docs/articles/viewing-epoch-based-metrics.rst:158]: (also generally, I have similar confusion why we are asking a user to "verify" code they downloaded form us)

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+       steps_completed=steps_completed,
+   )
+   scheduler.step()
+
+In addition, we'll need a ``steps_completed`` variable to plot metrics on a graph in the WebUI. The
+goal is to to track the progress of training by considering both the number of completed training
+batches and the current epoch index. This allows Determined to continuously plot the training loss
+on one graph for consecutive epochs.
+
+Verify your code contains a ``steps_completed`` variable:
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:184]: `steps_completed` is a mandatory parameter, so I don't think we need to write docs telling them to verify that it was passed... if they don't pass it, python itself will tell them it's a required parameter.

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+Your experiment opens in the **Overview** tab.
+
+-  Select the **Metrics** tab.
+-  Select the **X-Axis** menu and then choose **Epoch**.
+-  Scroll down to view the epoch-based metrics graph.
+
+.. image:: ../assets/images/webui-metrics-epoch-based.png
+   :width: 100%
+   :alt: Epoch-based metrics in the WebUI
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:214]: I like that you took this how-to all the way into what they would click in the webui.  I think that's something I've basically never thought to do.

File: docs/assets/images/webui-metrics-epoch-based.png
Comment (rb-determined-ai) [inline docs/assets/images/webui-metrics-epoch-based.png:1]: question: how will we remember to update this image when the webui changes its look and feel?

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+To sign in, we'll accept the default determined username, leave the password empty, and click **Sign
+In**.
+
+In the WebUI, we'll select our experiment, and then navigate to the **Logs** tab.
+
+************************
+ Step 2: Report Metrics
+************************
+
+In this section, we'll define our epoch metric. We'll also report training and validation metrics to
```
Comment (KevinMusgrave) [inline docs/articles/viewing-epoch-based-metrics.rst:81]: Maybe this could be shortened to just "In this section, we'll define our epoch metric."

And then "To follow along, use..."

File: docs/articles/viewing-epoch-based-metrics.rst (hunk 4 above)
Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst:110]: omitted

File: docs/articles/viewing-epoch-based-metrics.rst (hunk 2 above)
Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst]: dropped

File: docs/articles/viewing-epoch-based-metrics.rst (hunk 3 above)
Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst]: mitigated both comments by removing any steps that are already done in the downloaded scripts.

File: docs/articles/viewing-epoch-based-metrics.rst (hunk 5 above)
Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst]: mitigated as in previous comment

File: docs/articles/viewing-epoch-based-metrics.rst (hunk 6 above)
Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst]: mitigated both comments by removing any steps that are already done in the downloaded scripts.

File: docs/assets/images/webui-metrics-epoch-based.png
Comment (tara-det-ai) [reply docs/assets/images/webui-metrics-epoch-based.png:1]: i've moved this to a Jira ticket

File: docs/articles/viewing-epoch-based-metrics.rst (hunk 8 above)
Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst]: resolved

File: docs/articles/viewing-epoch-based-metrics.rst [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+However, we also want to report epoch-based metrics and to allow Determined to keep track of the
+specific epoch for which training loss is being reported.
+
+-  To do this, we'll modify the train() method to include ``epoch_idx`` as a metric:
+
+.. code:: python
+
+   # Index by (batch_idx + 1) * (epoch-1) * len(train_loader)
+   # to continuously plot loss on one graph for consecutive
+   # epochs.
```
Comment (rb-determined-ai) [inline docs/articles/viewing-epoch-based-metrics.rst:108]: Recommendation: drop this comment from the snippet.

As a developer, I usually read a comment first, because a comment indicates that the code is doing something non-obvious, or that it's a plain english explanation of what's going on.

But that comment is unhelpful, because it's drawing attention away from the line that was actually modified, which is the `metrics=` line.

Like, the comment makes sense from the perspective of somebody reading the original script, but it isn't the right focus from the perspective of this document.

Also, consider using `code:: diff` and using `-` in front of the line you're deleting and `+` in front of the line you are adding.  It will highlight the deleted line as red and the added line as green, in order to visually identify the modification to the script.  In fact, if you do that you could leave the comment if you'd like, because the dominant "focuser" would be the diff coloring, not the comment.

File: docs/_shared/note-local-dtrain-job.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+.. note::
+
+   This tutorial provides instructions for running a local distributed training job. Your setup may
+   be different. For example, for instructions on how to run a remote distributed training job,
+   visit the :ref:`qs-mdldev`.
```
Comment (rb-determined-ai) [inline docs/_shared/note-local-dtrain-job.txt:5]: I'm a bit confused about this, because I don't think the example in this code is distributed.

Can you elaborate a bit on what you're trying to communicate in this note?

Comment (tara-det-ai) [reply docs/_shared/note-local-dtrain-job.txt:5]: thanks for catching this. i've removed it.

File: docs/articles/viewing-epoch-based-metrics.rst (hunk 9 above)
Comment (tara-det-ai) [reply docs/articles/viewing-epoch-based-metrics.rst:108]: dropped the comment, thanks for the tip!
